### PR TITLE
Added full disk watcher and startup condition

### PR DESCRIPTION
**Pull request by berkayoz**
This would be a starting point to handle full disk cases. The conditions can checked by a watcher which will try to gracefully handover leadership/voter rights and shutdown. The syscall for `unix.Statfs` is really fast, so this check can be handled on queries that would put data to the disk such as `CREATE`,`UPDATE` and `DELETE`.  If having a watcher seems to be the better idea, this work can be finalized by cleaning up some of the ugly code.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -41,6 +41,18 @@ var (
 			if rootCmdOpts.debug {
 				logrus.SetLevel(logrus.TraceLevel)
 			}
+			var stat unix.Statfs_t
+
+			unix.Statfs(rootCmdOpts.dir, &stat)
+
+			avail := stat.Bavail * uint64(stat.Bsize)
+
+			var disk_threshold uint64 = 10000000
```
Comment (ktsakalozos) [inline cmd/root.go:50]: Could we have this configurable, as an argument. We should have also the query interval as an argument.

File: pkg/server/server.go [hunk 2]
```diff
@@ -249,6 +255,36 @@ func (s *Server) Start(ctx context.Context) error {
...
+				unix.Statfs(s.storageDir, &stat)
+				avail := stat.Bavail * uint64(stat.Bsize)
+				// If available space is less than 10MB reject to start.
+				if avail < disk_threshold {
+					// Create a separate context with 30 seconds to cleanup
+					stopCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+					defer cancel()
+
+					logrus.Debug("Handing over dqlite leadership")
+					if err := s.app.Handover(stopCtx); err != nil {
```
Comment (ktsakalozos) [inline pkg/server/server.go:272]: What happens if you are not a leader?

File: pkg/server/server.go [hunk 3]
```diff
@@ -37,6 +38,10 @@ var expectedFilesDuringInitialization = map[string]struct{}{
 	"tuning.yaml":    {},
 }
 
+var storageWatchTicker = time.NewTicker(5 * time.Second)
+var storageWatchQuit = make(chan struct{})
+var disk_threshold uint64 = 10000000
```
Comment (bschimke95) [inline pkg/server/server.go:43]: Let's reuse the value from above and not define it twice.

File: pkg/server/server.go [hunk 4]
```diff
@@ -16,13 +16,14 @@ import (
...
+	"golang.org/x/sys/unix"
 )
 
 // Server is the main k8s-dqlite server.
 type Server struct {
 	// app is the dqlite application driving the server.
-	app *app.App
-
+	app        *app.App
+	storageDir string
```
Comment (bschimke95) [inline pkg/server/server.go:26]: Please add a comment

File: pkg/server/server.go [hunk 5]
```diff
@@ -249,6 +255,36 @@ func (s *Server) Start(ctx context.Context) error {
 	}
 	logrus.WithFields(logrus.Fields{"id": s.app.ID(), "address": s.app.Address()}).Print("Started dqlite")
 
+	go func() {
+		for {
+			select {
+			case <-storageWatchTicker.C:
+				var stat unix.Statfs_t
+				unix.Statfs(s.storageDir, &stat)
```
Comment (bschimke95) [inline pkg/server/server.go:263]: You could move this logic to a util as this logic is also used in the other file.
e.g. 
```
if util.diskIsFull() {
...

File: pkg/server/server.go [hunk 6]
```diff
@@ -249,6 +255,36 @@ func (s *Server) Start(ctx context.Context) error {
...
+				avail := stat.Bavail * uint64(stat.Bsize)
+				// If available space is less than 10MB reject to start.
```
Comment (bschimke95) [inline pkg/server/server.go:265]: This comment will quickly run out of sync - IMHO better not add the concrete threshold here.

**Review by bschimke95 — COMMENTED** [review]
Nice work! Only a few comments.
Also, can we add a small test to cover that logic?

File: pkg/server/server.go (hunk 2 above)
Comment (berkayoz) [reply pkg/server/server.go]: Go-dqlite handles this in the background, only voter and leadership rights are handed over. Please see [app.go#L311-L316](https://github.com/canonical/go-dqlite/blob/9e965d50ce079bea63e8c76b18c07f3ca91fc2e4/app/app.go#L311-L316)

File: pkg/server/server.go [hunk 7]
```diff
@@ -237,18 +244,55 @@ func New(dir string, listen string, enableTLS bool, diskMode bool, clientSession
...
+					logrus.WithError(err).Errorf("failed to close dqlite app")
+				}
+				logrus.WithField("dir", s.swatcherConfig.storageDir).Fatalf("Disk is critically low, gracefully shutdown...")
+			}
+		case <-s.swatcherConfig.storageWatchQuit:
+			s.swatcherConfig.storageWatchTicker.Stop()
+			return
+		}
+	}
+}
```
Comment (neoaggelos) [inline pkg/server/server.go:285]: This should instead accept the `context.Context` passed in start, and return in `case <-ctx.Done()`

File: pkg/server/server.go [hunk 8]
```diff
@@ -237,18 +244,55 @@ func New(dir string, listen string, enableTLS bool, diskMode bool, clientSession
...
+
 // Start the dqlite node and the kine machinery.
 func (s *Server) Start(ctx context.Context) error {
 	if err := s.app.Ready(ctx); err != nil {
 		return fmt.Errorf("failed to start dqlite app: %w", err)
 	}
 	logrus.WithFields(logrus.Fields{"id": s.app.ID(), "address": s.app.Address()}).Print("Started dqlite")
 
+	go s.WatchStorage()
```
Comment (neoaggelos) [inline pkg/server/server.go:294]: Let's pass the context here

File: cmd/root.go [hunk 9]
```diff
@@ -42,6 +44,14 @@ var (
 				logrus.SetLevel(logrus.TraceLevel)
 			}
 
+			isFull, err := server.IsStorageFull(rootCmdOpts.dir, rootCmdOpts.storageDiskThreshold)
+			// If available space is less than threshold reject startup.
+			if err != nil {
+				logrus.WithError(err).Fatal("failed to check storage capacity")
+			} else if isFull {
+				logrus.WithField("dir", rootCmdOpts.dir).Fatal("Disk is critically low, rejecting startup")
```
Comment (neoaggelos) [inline cmd/root.go:52]:
```suggestion
				logrus.WithField("dir", rootCmdOpts.dir).Fatal("Disk is critically low, refusing to start")
```

File: pkg/server/server.go [hunk 10]
```diff
@@ -237,18 +244,55 @@ func New(dir string, listen string, enableTLS bool, diskMode bool, clientSession
...
+			storageDir:         dir,
+			storageWatchTicker: time.NewTicker(time.Duration(storageWatchPeriod) * time.Second),
+			storageWatchQuit:   make(chan struct{}),
+			diskThreshold:      storageDiskThreshold,
+		},
 		kineConfig: kineConfig,
 	}, nil
 }
 
+func (s *Server) WatchStorage() {
```
Comment (neoaggelos) [inline pkg/server/server.go:258]: This could be `watchStorage` instead, unless we need to export this?

File: pkg/server/server.go [hunk 11]
```diff
@@ -237,18 +244,55 @@ func New(dir string, listen string, enableTLS bool, diskMode bool, clientSession
...
+				defer cancel()
+				logrus.Debug("Handing over dqlite leadership")
+				if err := s.app.Handover(stopCtx); err != nil {
+					logrus.WithError(err).Errorf("Failed to handover dqlite")
+				}
+				logrus.Debug("Closing dqlite application")
+				if err := s.app.Close(); err != nil {
+					logrus.WithError(err).Errorf("failed to close dqlite app")
+				}
+				logrus.WithField("dir", s.swatcherConfig.storageDir).Fatalf("Disk is critically low, gracefully shutdown...")
```
Comment (neoaggelos) [inline pkg/server/server.go:278]: This should instead be handled by the context cancellation, we should not rewrite parts of `Shutdown()`.

I would recommend adding a `stopCh chan(struct{})` to the `Server` struct, then:

```
func (s *Server) Done() <-chan struct{} {
  return s.stopCh
}
```

And update the root command to wait for a cancellation like this:

```go
select {
case <-ch:  // received a signal
case <-server.Done(): // the server decided to terminate
}
```

**Review by neoaggelos — COMMENTED** [review]
Looks OK overall, leaving some comments to keep things tidy

**Pull request merged**